### PR TITLE
[OpenMP] Use generic IR for the OpenMP DeviceRTL

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
We previously built this for every single architecture to deal with
incompatibility. This patch updates it to use the 'generic' IR that
`libc` and other projects use. Who knows if this will have any
side-effects, probably worth testing more but it passes the tests I
expect to pass on my side.

### CODE REVIEW

File: offload/DeviceRTL/CMakeLists.txt
Comment (shiltian) [inline offload/DeviceRTL/CMakeLists.txt:126]: and nothing after `-march=`?

Comment (jhuber6) [reply offload/DeviceRTL/CMakeLists.txt:126]: Intentional

**Review by shiltian — COMMENTED** [review]
I like this method, but just out of curiosity, did we use anything in AMDGPU implementation that has target dependent lowering in the front end? If not, this is totally fine I'd say.

File: offload/DeviceRTL/CMakeLists.txt
Comment (shiltian) [reply offload/DeviceRTL/CMakeLists.txt:127]: so which means there is no arch?

**Comment by jhuber6** [comment]
> I like this method, but just out of curiosity, did we use anything in AMDGPU implementation that has target dependent lowering in the front end? If not, this is totally fine I'd say.

We used to use the `__AMDGCN_WAVEFRONT_SIZE` but that was removed for unrelated reasons.

File: offload/DeviceRTL/src/Reduction.cpp
Comment (shiltian) [inline offload/DeviceRTL/src/Reduction.cpp:88]: I'll try to make an AMDGPU counterpart for this one, though it doesn't look like necessary for the purpose of OpenMP device runtime.

Comment (jhuber6) [reply offload/DeviceRTL/src/Reduction.cpp:88]: I believe @AlexVlx also has interest in that space, so best consult with him as well.

**Comment by shiltian** [comment]
> > I like this method, but just out of curiosity, did we use anything in AMDGPU implementation that has target dependent lowering in the front end? If not, this is totally fine I'd say.
> 
> We used to use the `__AMDGCN_WAVEFRONT_SIZE` but that was removed for unrelated reasons.

And now we don't have any target specific part in AMDGPU implementation? Quite surprised but happy to hear that.

**Comment by shiltian** [comment]
Worth a bullet point in release note, both clang and openmp/offload.

**Comment by jhuber6** [comment]
@hidekisaito Might be relevant to your patch.

**Review by shiltian — APPROVED** [review]
it seems like there are some test failures but I think the PR looks good

**Comment by jhuber6** [comment]
@ronlieb This will probably conflict a lot with downstream, should probably wait until I've talked it over with others before trying to merge it in AOMP.

**Comment by ronlieb** [comment]
> @ronlieb This will probably conflict a lot with downstream, should probably wait until I've talked it over with others before trying to merge it in AOMP.

yes please , lets wait until internal monday mtg to discuss. also i would like to help with some precommit performance testing   using a downstream build.

File: offload/DeviceRTL/CMakeLists.txt
Comment (mgorny) [inline offload/DeviceRTL/CMakeLists.txt:265]: Can we have an option to disable either of amdgpu/nvptx specifically? If anything, because it feels weird having a choice between amdgpu and nvptx plugins, but no choice between devicertl variants.

Comment (jhuber6) [reply offload/DeviceRTL/CMakeLists.txt:265]: I didn't feel like it was strictly necessary since if you can build one you can build the other. The only thing you save is disk space, but if you feel like it's really needed I can re-use the architectures thing or something.

Comment (mgorny) [reply offload/DeviceRTL/CMakeLists.txt:265]: I don't exactly feel strongly about it. It's just that we already provide a switch for amdgpu/nvptx in Gentoo, so it only feels natural for this to respect it.

Also, on a semi-related matter: is there a reason we're installing both separate `.bc` files and `libomptarget.devicertl.a`?

Comment (jhuber6) [reply offload/DeviceRTL/CMakeLists.txt:265]: This is really over-complicated, but right now we only use the `.bc` file for non-LTO NVPTX compilations which get put through each TU in a broken way via `-mlink-builtin-bitcode`. This is because it would get really slow if we didn't optimize out calls to the runtime.

Comment (mgorny) [reply offload/DeviceRTL/CMakeLists.txt:265]: Ok, so it's not accidental/obsolete — that's all I needed to know :-).

Another thing I've noticed that clang needs `libomptarget.devicertl.a` even when no GPU target is actually used. I've worked around that for now by creating an empty `.a` file locally, but perhaps that should also be handled better somehow (I can make a PR later if you tell me how you'd like it handled).

Comment (jhuber6) [reply offload/DeviceRTL/CMakeLists.txt:265]: I suppose that happens for the CPU  targets? Those aren't really used anywhere real, they're just for testing, but even so we probably shouldn't leave them broken. We link the device RTL unconditionally because the link step shouldn't need to know the architectures that were used to compile it. Because these are static libraries they're not extracted if they aren't needed, so it doesn't hurt anything if they're unused. But I guess you can have a situation where the user doesn't build this intentionally, but that's non-default behavior so I've never given it much thought.

**Comment by arsenm** [comment]
This probably should retain separate wave32/wave64 builds. Additionally, should have extension points for subtarget specific implementations

**Comment by jhuber6** [comment]
> This probably should retain separate wave32/wave64 builds. Additionally, should have extension points for subtarget specific implementations

That's what Shilei was talking about since we have `__nvvm_reflect` for that for NVPTX.

**Comment by arsenm** [comment]
> > This probably should retain separate wave32/wave64 builds. Additionally, should have extension points for subtarget specific implementations
> 
> That's what Shilei was talking about since we have `__nvvm_reflect` for that for NVPTX.

First I will always consider NVVM reflect a giant hack. NVVM reflect cannot actually deal with the full range of wavesize issues. It is an incompatible ABI and the code should never be intermixed

**Comment by shiltian** [comment]
> This probably should retain separate wave32/wave64 builds. Additionally, should have extension points for subtarget specific implementations

Does wave32/wave64 affect the lowering from front end source code to middle end IR, if we don't use certain functions, such as wave level primitives?

File: offload/DeviceRTL/CMakeLists.txt
Comment (mgorny) [reply offload/DeviceRTL/CMakeLists.txt:265]: Yeah. precisely. I've noticed this while building `-DLIBOMPTARGET_BUILD_DEVICERTL_BCLIB=OFF -DLIBOMPTARGET_PLUGINS_TO_BUILD=host` to run tests on CPU only.

**Comment by arsenm** [comment]
> Does wave32/wave64 affect the lowering from front end source code to middle end IR, if we don't use certain functions, such as wave level primitives?

Yes. This is a fixed ABI setting that can change the binary function signature (which then has knock on optimization effects). The IR is a representation of what goes on a final .o. It is not really intended to enable new features or make the program portable.

**Comment by shiltian** [comment]
> The IR is a representation of what goes on a final .o. It is not really intended to enable new features or make the program portable.

My .02 is, it depends on how to think about it. If we eliminate the "unwanted" part (especially those that could potentially affect ABI) at the very beginning of the middle end, it will not have optimization effects.

In addition, the compiler backend lowering is essentially doing something like:

```
if (target == A)
  lower Inst to I1
else if (has_feature_that_only_compiler_can_call("..."))
  lower Inst to I2
...
```

If speaking from "portability", the compiler is "portable". The reflection is just to give end user an interface to do similar things:

```
if (__builtin_get_target() == A)
  do something in one way
else if (has_feature_that_user_can_call("..."))
  do something in another way
...
```

Of course the interface(s) have to be carefully designed and all the queried information can't be changed during optimization.

**Review by thesamesam — APPROVED** [review]
LGTM (thanks for pinging mgorny and I), notwithstanding mgorny's comment, but we could handle that as a followup too.

**Comment by mgorny** [comment]
@jhuber6, since I think this PR makes #119006 obsolete, should I merge that one while this one's still in the works, or close it?

**Comment by jhuber6** [comment]
> @jhuber6, since I think this PR makes #119006 obsolete, should I merge that one while this one's still in the works, or close it?

I wouldn't say it's high priority. I would've landed this already if it weren't for the downstream fork being so divergent this isn't an easy apply.

**Comment by jhuber6** [comment]
@ronlieb https://gist.github.com/jhuber6/26e3fc8473a750410d2ca1db4a65918d

**Review by ronlieb — APPROVED** [review]

**Pull request merged**
